### PR TITLE
Andrey Rohau / Added Creation Functionality for CreateTestCommand

### PR DESCRIPTION
**Pull request by andreyrogov8**

### CODE REVIEW

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 1]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+                CreateTestAuditionQuestions(test, cancelationtoken);
+
                 _context.Tests.Add(test);
                 await _context.SaveChangesAsync(cancelationtoken);
 
                 return new Response { Id = test.Id };
             }
+
+            public Guid SetEssay(IQueryable<Topic> topics)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:59]: Why public?

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 2]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+            }
+
+            public Guid SetAudition(IQueryable<Audition> auditions)
+            {
+                return auditions
+                    .ElementAt(_rand.Next(0, auditions.Count()))
+                    .Id;
+            }
+
+            public async void CreateTestGrammarQuestions(Test test, CancellationToken cancelationtoken)
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:81]: add Async postfix please

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 3]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+            {
+                var questions = _context
+                    .Questions
+                    .AsNoTracking()
+                    .Where(x => x.Level == test.Level && x.AuditionId == null).Take(12);
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:86]: but here are 12 first questions always

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 4]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+                foreach (var question in questions)
+                {
+                    var testQuestion = new TestQuestion()
+                    {
+                        Id = Guid.NewGuid(),
+                        TestId = test.Id,
+                        QuestionId = question.Id
+                    };
+                    _context.TestQuestions.Add(testQuestion);
+                    await _context.SaveChangesAsync(cancelationtoken);
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:97]: you should save once for whole handler, at least not in for each

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 5]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+            }
+
+            public async void CreateTestAuditionQuestions(Test test, CancellationToken cancelationtoken)
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:101]: add Async postfix please

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 6]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
             }
+
+            public Guid SetEssay(IQueryable<Topic> topics)
+            {
+                return topics
+                    .ElementAt(_rand.Next(0, topics.Count()))
+                    .Id;
+            }
+
+            public Guid SetSpeaking(IQueryable<Topic> topics, Test test)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:66]: instead test you can pass just EssayId as a parameter IMHO

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 7]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+                    await _context.SaveChangesAsync(cancelationtoken);
+                }
+            }
+
+            public async void CreateTestAuditionQuestions(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = _context
+                    .Questions
+                    .AsNoTracking()
+                    .Where(x => x.Level == test.Level && x.AuditionId == test.AuditionId).Take(10);
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:106]: here first 10 questions, the same for each time

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 6 above)
Comment (DShcharbakou) [reply Elevel/Application/Features/TestCommands/CreateTestCommand.cs:66]: Why public?

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 8]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+                foreach (var question in questions)
+                {
+                    var testQuestion = new TestQuestion()
+                    {
+                        Id = Guid.NewGuid(),
+                        TestId = test.Id,
+                        QuestionId = question.Id
+                    };
+                    _context.TestQuestions.Add(testQuestion);
```
Comment (parys) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:116]: same, ideally to call Save just once at the end of handler, not each object

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 9]
```diff
@@ -24,20 +26,97 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+            public Guid SetSpeaking(IQueryable<Topic> topics, Test test)
+            {
+                return topics
+                    .Where(x => x.Id != test.EssayId)
+                    .ElementAt(_rand.Next(0, topics.Count()))
+                    .Id;
+            }
+
+            public Guid SetAudition(IQueryable<Audition> auditions)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:74]: Why public?

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 2 above)
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:81]: Why public?

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 3 above)
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:86]: What is it (12)? magic number

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 5 above)
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:101]: Why public?

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 7 above)
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:106]: What is it (10)? magic number

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

**Pull request closed without merging**